doc: Improve getting started #2606

francisu opened the pull request:
Ready to merge from my perspective. All comments have been addressed. More work can be done, but will be in a separate PR.

htuch commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+Getting Started
+===============
+
+This section gets you started with a very simple configuration, and provides some example configurations.
```
Prefer no comma in this situation (doesn't seem to be Oxford comma as there aren't 3 items).

htuch commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+Envoy can be configured using a single YAML file passed in as an argument on the command line.
+
+The :ref:`admin message <envoy_api_msg_config.bootstrap.v2.Admin>` is required to configure
+the administration server. The `address` key specifies the :ref:`address <envoy_api_file_envoy/api/v2/core/address.proto>`
```
"specifies the listening address"

htuch commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+
+The :ref:`admin message <envoy_api_msg_config.bootstrap.v2.Admin>` is required to configure
+the administration server. The `address` key specifies the :ref:`address <envoy_api_file_envoy/api/v2/core/address.proto>`
+which in this case is simply `localhost:9901`.
```
Technically its `0.0.0.0:990`, since `localhost` might resolve to `::1` (or anything you put in `/etc/host`).

htuch commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+      socket_address: { address: 0.0.0.0, port_value: 9901 }
+
+The :ref:`static_resources <envoy_api_field_config.bootstrap.v2.Bootstrap.static_resources>` contains everything that is configured staticaly when Envoy starts,
+as opposed to the means of configuring resources dynamically when Envoy is running.
```
Internal link to https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/v2_overview

htuch commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+        - filters:
+          - name: envoy.http_connection_manager
+            config:
+              stat_prefix: ingress_http
```
Do we want a comment about each field here describing what is roughly meant?

ccaraman replied:
I think that will be very helpful. Ex this is the http l7 filter that can provide and for routing domain ‘*’ will match all requests on on this port.

francisu replied:
Yes, this is part of the WIP, I did not finish this section, and I will do it at a detailed level going through each field (or collection of fields, as it makes sense).

htuch commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+
+    static_resources:
+
+The :ref:`listeners <envoy_api_file_envoy/api/v2/listener/listener.proto>` blah.
```
I take it that these "blahs" are WIP markers :)

francisu replied:
Yup, I will greatly expand these sections.

ccaraman commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+  FROM envoyproxy/envoy:latest
+  RUN apt-get update
+  COPY envoy.yaml /etc/envoy.yaml
+  CMD /usr/local/bin/envoy -c /etc/envoy.yaml
```
Add a link to the command line documents.

francisu replied:
Done

ccaraman commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+feature of the API, which is most useful for simple requirements. For more complex requirements
+:ref:`Dynamic Configuration <arch_overview_dynamic_config>` is supported.
+
+Quick Start to Run Simple Example
```
Nit: this file will grow pretty quickly. It might make more sense to have each section in its own file.

francisu replied:
Sure, could we do that when it expands?

htuch replied:
It's up to you. If you just want to land this PR to get going and plan on following up with two other PRs for splitting up the content and expanding the "Simple Configuration" section, that's fine, I agree with the idea of breaking this up, since the "Simple Configuration" is distinct to the later bits on sandboxes. If you'd like to do that and think this PR is ready for final review, can give this another pass.

francisu replied:
@htuch Yes, I would like to get this one in quickly, as it touches a lot of files, and involves a top-level reorg. Subsequent work will be only in a couple of files and thus easier to maintain a PR. I think it's ready for final review.

ccaraman commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+request to port 10000::
+
+  $ docker build -t envoy-google-test:v1 .
+  $ docker run -d -p 10000:10000 envoy-google-test:v1
```
Nit: is there a link in here to docker and how to set it up locally?

ccaraman replied:
And can each command be commented

francisu replied:
The each command is separated and commented on in the section below. This section was essentially the older quickstart (and I added a paragraph in the front to say the explanation is given below), so I think this is OK (would you agree?).  I will add the docker link though.

francisu replied:
On second thought, in general I don't see documentation linking to Docker instructions, as they are outside of the scope of the documentation, and we should not have links to places where things could change. So I think we can assume people know how to deal with Docker and if they can't, they are able to operate teh google to figure it out.

ccaraman commented on docs/root/start/start.rst:
```diff
@@ -0,0 +1,154 @@
+the administration server. The `address` key specifies the :ref:`address <envoy_api_file_envoy/api/v2/core/address.proto>`
+which in this case is simply `localhost:9901`.
+
+.. code-block:: yaml
```
Please add a link to the admin interface docs. “operations_admin_interface”

francisu commented:
What I would like to do is get this one in pretty much as it is (addressing the above comments and removing the "blah" text, but not finishing any more detail). Because it's a bit of a reorganization, putting it in sooner than later will be easier. And then I can do another PR to flesh out the more detailed explanation. @htuch @ccaraman do you think that's a good approach. I can address Carmen's comments now, and then hopefully it can be merged. Assuming @richarddli is OK with the organizational changes -- I think though further refinement on the organization is necessary as the stuff that points to the Datawire stuff might want a better home for example; but this is a step in the right direction.

richarddli commented:
I like the idea of creating a separate Getting Started section, and promoting it to the top level out of the Building/Installation section.

I'm trying to understand what we want the end user to accomplish at the end of the "Simple Configuration" section.  Reading the text right now, it seems it:

1. Is doing an overview of the config file format. I think we want to limit how much we do here, since this should be covered elsewhere in the documentation.
2. talks about how to set up the administration server. This is useful, and is an obvious way to build on the previous example. 

If we can craft a Getting Started that incrementally layers on tutorials/lessons as you walk through it, I think that simplifies the learning curve. So my suggestion would be something like:

1. Quick start -- use Envoy to route to Google
2. Then configure admin server
3. Then add statsd
4. Then bundle all this configuration into a custom Docker image ready for production

Also, if we do this, you may want to pull the Simple Configuration section into a separate document, instead of making it one big flow.

francisu commented:
@richarddli The intention of the simple configuration section is to be much larger, as it will go through all of the parameters in detail. It's only short because I did not get to that work. I wanted to give an expanded tutorial as an alternative to the previous text that just had you take the files and run them (and I accept that the intent of that is very unclear from the text).

I guess how I think of this is that it would be more or less tutorials for doing the simple things, with a lot of links referring to the reference. The previous text using the files could be put at the end of the tutorial, with a link to it at the beginning (so you can skip the tutorial if you want).

So the Simple Configuration will be a tutorial which will be all you need to get a basic understanding and deploy Envoy for simple cases (I agree it can be a separate document). And we may have others that are like it with more specific configurations.

richarddli commented:
Got it. So, after completing the Simple Configuration tutorial, what would the user accomplish, other than having a better understanding of the config file format?

francisu commented:
@richarddli. I would expect them to know enough to make a docker-based deployment of a simple proxy and have familiarity with the API reference enough to do other things. Perhaps there can be follow on tutorials that build on that for other cases. Augmenting the sandboxes (and being coordinated with them)

richarddli commented:
Could we do something more specific, such as configuring another listener to talk to `httpbin.org`, or adjusting timeouts, or something like that? Otherwise, it sounds like it's just a walk-through and not a tutorial.

francisu commented:
Yes. I agree. Some things that are actually useful. I will prepare something. But I would like to get this pr merged and build on top of that.

richarddli commented:
@mattklein123 To clarify, I'm not disagreeing with the "better understand config file format" goal. I just think if we're putting it here, tutorial-style, then we should make it a hands-on exercise where you do something useful, with a focus on communicating the mental model of the config format.

@francisu sounds like you agree that you want to do something useful in the tutorial in subsequent PRs, and just focus this PR on structure, which I think is fine. If you're focusing this on structure, do you want to go ahead and create the "Simple Configuration" file / chapter and do the pull out now?
